core/tests/puppeteer-acceptance-tests/puppeteer-testing-utilities/puppeteer_utils.js

```diff
@@ -0,0 +1,240 @@
+   */
+  async signInWithEmail(email) {
+    await this.goto(testConstants.URLs.home);
+    if (!this.userHasAcceptedCookies) {
+      await this.clickOn('button', 'OK');
+      this.userHasAcceptedCookies = true;
+    }
+    await this.clickOn('span', 'Sign in');
+    await this.type(testConstants.SignInDetails.inputField, email);
+    await this.clickOn('span', 'Sign In');
+    await (this.page).waitForNavigation({waitUntil: 'networkidle0'});
+  }
+
+  /**
+   * This function signs up a new user with the given username and email.
+   * @param {string} userName - The username of the user.
+   * @param {string} signInEmail - The email of the user.
+   */
+  async signUpNewUser(userName, signInEmail) {
+    await this.signInWithEmail(signInEmail);
+    await this.type('input.e2e-test-username-input', userName);
+    await this.clickOn('input', 'e2e-test-agree-to-terms-checkbox');
+    await this.page.waitForSelector(
+      'button.e2e-test-register-user:not([disabled])');
+    await this.clickOn('button', 'Submit and start contributing');
+    await (this.page).waitForNavigation({waitUntil: 'networkidle0'});
+  }
+
+  /**
+   * This function waits for a component using its CSS selector to load.
+   * @param {string} selector - The CSS selector of the component.
+   */
+  async waitForPageToLoad(selector) {
+    await (this.page).waitForSelector(selector);
+  }
+
+  /**
+   * This function reloads the current page.
+   */
+  async reloadPage() {
+    await (this.page).reload({waitUntil: ['networkidle0', 'domcontentloaded']});
+  }
+
+  /**
+   * This function clicks on any element using its CSS selector
+   * or the text written on that element.
+   * @param {string} tag - The HTML tag of the element.
+   * @param {string} selector - The CSS selector or the text
+   * written on the element.
+   */
+  async clickOn(tag, selector) {
+    try {
+      await (this.page).waitForXPath('//' + tag);
+      const [button] = await (this.page).$x(
+        '//' + tag + '[contains(text(), "' + selector + '")]');
+      await button.click();
+    } catch {
+      await (this.page).waitForSelector(tag + '.' + selector);
+      await (this.page).click(tag + '.' + selector);
+    }
+  }
+
+  /**
+   *
+   * @param {string} username - The username to which role would be assigned.
+   * @param {string role - The role that would be assigned to the user.
+   */
+  async assignRoleToUser(username, role) {
+    await this.goto(rolesEditorTab);
+    await this.type(roleEditorInputField, username);
+    await this.clickOn('button', roleEditorButtonSelector);
+    await this.clickOn('h4', 'Add role');
+    await this.clickOn('div', rolesSelectDropdown);
+    await this.page.evaluate(async(role) => {
+      const allRoles = document.getElementsByClassName('mat-option-text');
+      for (let i = 0; i < allRoles.length; i++) {
+        if (allRoles[i].innerText.toLowerCase() === role) {
+          allRoles[i].click({waitUntil: 'networkidle0'});
```

Comment:
Can you add a comment about what does `{waitUntil: 'networkidle0'}` mean?

Reply:
ditto elsewhere

Reply:
this thing is waiting until the network request is 0 for 500ms implying the loading is finished. this is required so the navigation is completed. for example. this waituntil networkidle0 is also used in goto function so that next test steps should wait until the page is fully loaded after the navigation or network request is completed. will be adding a JS Doc to explain this more in the file.